### DISCUSSION TITLE
getting "starlette.middleware.exceptions:ExceptionMiddleware.__call__" in API

### DISCUSSION THREAD

**Discussion by harshit-2115** [fastapi/fastapi#12416] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_headers=["*"],
    allow_methods=["GET"]
)

@app.get("/predict")
def predict(query: str):
    try:
        result = customFunc()
        if(result == "x"):
            raise HTTPException(status_code=406, detail=result)    
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail="Server Error")
```

### Description

<img width="1331" alt="Screenshot 2024-10-09 at 7 28 10 PM" src="https://github.com/user-attachments/assets/3c31109b-a3cf-41ae-96a9-af45d05a5965">

Hello

Could anyone help me with this exception call I'm getting **starlette.middleware.exceptions:ExceptionMiddleware.__call__**

Is it because I don't have an exception handler to handle the HTTPException that I'm raising in my code ?

Other than CORS, no other middle is used.

### Operating System

Linux

### FastAPI Version

0.110.1

### Pydantic Version

2.6.4

### Python Version

3.9.11

**Comment by sinisaos** [comment]
@harshit-2115 I don't know if it helps, but every Starlette application (which FastAPI actually is) automatically includes `ServerErrorMiddleware` and `ExceptionMiddleware` by default. [Here](https://www.starlette.io/middleware/) is the documentation. Maybe [this discussion](https://github.com/fastapi/fastapi/issues/4241) can help in your case

**Reply by harshit-2115** [reply]
Thanks for the response. 

In that discussion you shared, advise is to change from async def to just a def. In my case, I'm already using def, no async/await. 

Can you please tag someone who might help here ?

**Reply by sinisaos** [reply]
@harshit-2115 Unfortunately not. I can only add @Kludex which seems to be the most active.

**Reply by Kludex** [reply, marked as answer]
https://www.starlette.io/middleware/#using-middleware

<img width="734" alt="Screenshot 2024-10-10 at 21 10 17" src="https://github.com/user-attachments/assets/a82fdda2-7466-467c-a957-688821bd9a93">

The `ExceptionMiddleware` wraps the endpoint, and other stuff.
> In that discussion you shared, advise is to change from async def to just a def. In my case, I'm already using def, no async/await.

That depends on what `customFunc` does.

In any case, no issue about the `ExceptionMiddleware`.

**Reply by harshit-2115** [reply]
> The ExceptionMiddleware wraps the endpoint, and other stuff.

Please refer to the image I shared, **Avg Calls / Txn** is 1 for this Exception handler, so this trace is affecting all the requests.

Being a exception handler, shouldn't it just affect those requests where HTTPException is being raised ?

**Comment by Kludex** [comment]
@harshit-2115 Can you try and check if this PR: https://github.com/encode/starlette/pull/2727 improves something? 

You can install the branch via:

`pip install git+https://github.com/encode/starlette.git@move-read-of-starlette.exceptions-handler-when-exception-happens`
